### PR TITLE
Bug fix to getShortestPath

### PR DESCRIPTION
**Pull request by JLVarjo**
There was issue in getShortestPath with fused rings and certain atom indexing. Briefly, the problem was that the "pred" list was not checked for earlier insertions, causing the possible shorter paths to be overwritten later. I also cleaned up the code a bit: the information carried in "doneAtms" list is now merged to the "pred" list and doneAtms is no longer needed.

To reproduce the issue, run the added test in molopstest.cpp and see how the path does not go through the ring junctions as it should, leading to path length of 8 atoms, instead of 7.

### CODE REVIEW

File: Code/GraphMol/Matrices.cpp [hunk 1]
```diff
@@ -331,16 +331,10 @@ INT_LIST getShortestPath(const ROMol &mol, int aid1, int aid2) {
...
-
-  // pred.reserve(nats);
-  // doneAtms.reserve(nats);
-  pred.resize(nats);
-  doneAtms.resize(nats);
-  int ai;
-  for (ai = 0; ai < nats; ++ai) {
-    doneAtms[ai] = 0;
-  }
+  INT_VECT pred;
```
Comment (greglandrum) [inline Code/GraphMol/Matrices.cpp]: These two lines should just be:`INT_VECT pred(nats,-1);`

File: Code/GraphMol/Matrices.cpp [hunk 2]
```diff
@@ -352,20 +346,25 @@ INT_LIST getShortestPath(const ROMol &mol, int aid1, int aid2) {
...
-        if (static_cast<int>(*nbrIdx) == aid2) {
+      switch (pred[*nbrIdx]) {
+        case -1:
+          pred[*nbrIdx]=curAid;
+          bfsQ.push_back(rdcast<int>(*nbrIdx));
+          break;
+        case -3: // end found
+          pred[*nbrIdx] = curAid;
           done = true;
+          goto EXIT_LOOP;
```
Comment (greglandrum) [inline Code/GraphMol/Matrices.cpp]: The `goto` isn't necessary if you change line 348 to `while(!done && nbrIdx != endNbrs) {`

File: Code/GraphMol/Matrices.cpp [hunk 3]
```diff
@@ -352,20 +346,25 @@ INT_LIST getShortestPath(const ROMol &mol, int aid1, int aid2) {
...
+          break;
+        default: // already processed (or begin)
           break;
         }
-        bfsQ.push_back(rdcast<int>(*nbrIdx));
-      }
       nbrIdx++;
```
Comment (greglandrum) [inline Code/GraphMol/Matrices.cpp]: This is bad code that was already there, but it would be great if you could go ahead and change this to `++nbrIdx`

**Comment by greglandrum** [comment]
Thanks for catching and fixing this. I made a couple of small suggestions. I think the code is ready to merge after those are addressed.

**Comment by JLVarjo** [comment]
Sorry about the hassle with the non-source files, should be ok now...

**Comment by greglandrum** [comment]
Thanks !

**Pull request merged**